lib/spack/spack/test/database.py

```diff
@@ -17,13 +17,171 @@
 
 import spack.repo
 import spack.store
-from spack.test.conftest import MockPackageMultiRepo
+import spack.database
+import spack.spec
+from spack.test.conftest import MockPackage, MockPackageMultiRepo
 from spack.util.executable import Executable
 
 
 pytestmark = pytest.mark.db
 
 
+class MockLayout(object):
+    def __init__(self, root):
+        self.root = root
+
+    def path_for_spec(self, spec):
+        return self.root + '/'.join(['test_upstream_root', spec.name])
+
+    def check_installed(self, spec):
+        return True
+
+
+@pytest.fixture()
+def test_store(tmpdir):
+    real_store = spack.store.store
+    spack.store.store = spack.store.Store(str(tmpdir.join('test_store')))
+
+    yield
+
+    spack.store.store = real_store
+
+
+@pytest.mark.usefixtures('config')
+def test_installed_upstream(tmpdir_factory, test_store):
+    mock_db_root = str(tmpdir_factory.mktemp('mock_db_root'))
+    prepared_db = spack.database.Database(mock_db_root)
+
+    # Generate initial DB file to avoid reindex
+    with open(prepared_db._index_path, 'w') as db_file:
+        prepared_db._write_to_file(db_file)
+
+    default = ('build', 'link')
+    x = MockPackage('x', [], [])
+    z = MockPackage('z', [], [])
+    y = MockPackage('y', [z], [default])
+    w = MockPackage('w', [x, y], [default, default])
+    mock_repo = MockPackageMultiRepo([w, x, y, z])
+
+    mock_layout = MockLayout('/a/')
```

Comment:
I think this should use tmpdir for the layout root. I don't think this should write to the layout, but if something goes wrong we don't want to introduce side-effects.